versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/CommitLogicImpl.java

```diff
@@ -0,0 +1,1094 @@
+          }
+          if (n.isEmpty()) {
+            return endOfData();
+          }
+
+          try {
+            b = batch = Arrays.asList(persist.fetchObjs(n.toArray(new ObjId[0]))).iterator();
+          } catch (ObjNotFoundException e) {
+            throw new NoSuchElementException(
+                "Commit(s) "
+                    + e.objIds().stream().map(ObjId::toString).collect(Collectors.joining(", "))
+                    + " not found");
+          }
+        }
+
+        if (b.hasNext()) {
+          CommitObj c = (CommitObj) b.next();
+
+          if (c == null) {
+            // oops, commit not found...
+            return endOfData();
+          }
+
+          if (c.id().equals(endCommitId)) {
+            batch = emptyIterator();
+            next = null;
+          } else if (!b.hasNext()) {
+            next = c.tail();
+          }
+
+          return c;
+        }
+      }
+    }
+
+    @Nonnull
+    @jakarta.annotation.Nonnull
+    @Override
+    public PagingToken tokenForKey(ObjId key) {
+      return key != null ? pagingToken(key.asBytes()) : emptyPagingToken();
+    }
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public PagedResult<ObjId, ObjId> commitIdLog(
+      @Nonnull @jakarta.annotation.Nonnull CommitLogQuery commitLogQuery) {
+    ObjId startCommitId =
+        commitLogQuery
+            .pagingToken()
+            .map(PagingToken::token)
+            .map(ObjId::objIdFromBytes)
+            .orElse(commitLogQuery.commitId());
+
+    return new CommitIdIter(startCommitId, commitLogQuery.endCommitId().orElse(null));
+  }
+
+  private final class CommitIdIter extends AbstractIterator<ObjId>
+      implements PagedResult<ObjId, ObjId> {
+    private final ObjId endCommitId;
+
+    private Iterator<ObjId> batch;
+    private List<ObjId> next;
+
+    CommitIdIter(ObjId startCommitId, ObjId endCommitId) {
+      this.next = singletonList(startCommitId);
+      this.endCommitId = endCommitId;
+    }
+
+    @Override
+    protected ObjId computeNext() {
+      while (true) {
+        Iterator<ObjId> b = batch;
+        if (b == null || !b.hasNext()) {
+          List<ObjId> n = next;
+          next = null;
+
+          if (n == null) {
+            return endOfData();
+          }
+          int i = n.indexOf(EMPTY_OBJ_ID);
+          if (i != -1) {
+            n = n.subList(0, i);
+          }
+          if (n.isEmpty()) {
+            return endOfData();
+          }
+
+          b = batch = n.iterator();
+        }
+
+        if (b.hasNext()) {
+          ObjId c = b.next();
+
+          if (c.equals(endCommitId)) {
+            batch = emptyIterator();
+            next = null;
+          } else if (!b.hasNext()) {
+            CommitObj obj;
+            try {
+              obj = fetchCommit(c);
+            } catch (ObjNotFoundException e) {
+              throw new NoSuchElementException("Commit '" + c + "' not found");
+            }
+            if (obj == null) {
```

Comment:
Updated `Persist.fetchObj*` function to not return `null`.

Just not happy with having a non-null `CommitObj` for `EMPTY_OBJ_ID`, because that would "materialize" the "nothing commit" and require us to update the logic to handle empty parent commit list.

Reply:
fair enough :+1: